File: src/test/python/parser-opentrackIO/opentrackIOlib.py
# opentrackIO_parser.py
#
# Reference code for decoding opentrackIO messages
# Copyright (c) 2024 Steve Rosenbluth, RiS OSVP camera tracking committee
#
# License: this code is open-source under the FreeBSD License

import json
import argparse
import os

# Class to decode and interpret the protocol
# Pass in either a JSON message or a test filename
class OTProtocol:
    def __init__(self, msg_str=None, filepath=None):
        # msg_str = string containing a single json "sample"
        # filepath: file containing a JSON "sample", for testing
        self.sample_str = msg_str
        self.filepath = filepath
        self.schemapath = None
        self.trans_mult = 1.0                               # user-preferred units multiplier
        self.curtime = 0
        self.cam_trans = {"X":0,"Y":0,"Z":0}
        self.pd = None          # the parsed procotol dictionary
        self.sd = None          # the parsed schema dictionary
        self.sample_time_format = "seconds"
        self.sample_time_formats = ["seconds","timecode","string"]
        self.verbose = False
        self.focus_dist_mult = 1.0

        parser = argparse.ArgumentParser(description='OpenTrackingProtocol parser')
        parser.add_argument('-f', '--file', help='The JSON input file.', default=None)
        parser.add_argument('-s', '--schema', help='The schema (JSON) input file.', default=None)
        parser.add_argument('-t', '--test', help='If present, will only verify json',action='store_true')
        parser.add_argument('-v', '--verbose', help='Make script more verbose',action='store_true')
        args = parser.parse_args()
        if (args.schema):
            if os.path.exists(args.schema):
                self.schemapath = args.schema
        if self.schemapath:
            self.Import_schema()
        if (args.file):
            if os.path.exists(args.file):
                self.filepath = args.file
        if (args.verbose):
            self.verbose = True

        self.Parse()                # parse the actual JSON of the protocol

    def Import_schema(self):
        text = ''
        try:
            if (self.schemapath != None):        
                with open(self.schemapath, 'r') as fd:
                    print("Reading OTIO schema file: {0}".format(self.schemapath))
                    lines = fd.readlines()
                    for line in lines:
                        text = text + line
            self.sd = json.loads(text)
        except json.decoder.JSONDecodeError:
            print("Got JSONDecodeError decoding the sample!")
            self.sd = None
        if not self.sd:
            print("Parse(): Failed to parse OTIO schema file.")
        else:                                   # we have a OTP message
            print("Parsed the schema JSON successfully.")
            if self.verbose:
                print("Contents of the parsed JSON schema dict:\n")
                print(self.sd)
                print("\n")
            #print("camera is: {}".format(self.sd["properties"]["camera"]))
            #print("finding: {}".format(self.sd["properties"]["lens"]["properties"]["focalLength"]["units"]))

    # ingest the text and store the JSON items in a dictionary
    def Parse(self):        
        protocol = None
        text = ''
        if (self.filepath != None):        
            with open(self.filepath, 'r') as fd:
                print("Reading OTIO sample file: {0}".format(self.filepath))
                lines = fd.readlines()
                for line in lines:
                    text = text + line
        elif self.sample_str != None:        
            print("Parsing JSON string from sample buffer...")
            text = self.sample_str
        #print(text)
        #print('\n')    
        try:
            self.pd = json.loads(text)
        except:
            print("Error: failed to parse JSON file! Either path incorrect or JSON invalid.")
            exit(-1)
        if not self.pd:
            print("Parse(): Failed to parse OTIO message file.")
        else:                                   # we have a OTP message
            print("Parsed the sample OTIO JSON successfully.")
            if self.verbose:
                print("Contents of the parsed JSON dict:\n")
                print(self.pd)
                print("\n")

    # Returns the multiplicative conversion factor from meters to user-specified units
    # Valid unit specifier strings: "m", "cm", "mm", "in"
    def Conversion_factor_from_meters(self,unit_str):
        if unit_str == "m":
            return 1.0
        elif unit_str == "cm":
            return 100.0
        elif unit_str == "mm":
            return 1000.0
        elif unit_str == "in":
            return (1000/25.4)

    def Get_camera_trans(self, dimension,object_name=None):
        for tr in self.pd["transforms"]:
            if ("Camera" in tr["name"]):
                if (dimension == 'X'):
                    return tr["translation"]["x"] * self.trans_mult
                elif (dimension == 'Y'):
                    return tr["translation"]["y"] * self.trans_mult
                elif (dimension == 'Z'):
                    return tr["translation"]["z"] * self.trans_mult
                break

    def Get_camera_translations(self,object_name=None):
        return (self.Get_trans('X'), self.Get_trans('Y'), self.Get_trans('Z'))

    def Get_timecode(self):
        if ("timing" in self.pd) and (self.pd["timing"]["timecode"]):
            hh = '{:02}'.format(int(self.pd["timing"]["timecode"]["hours"]))
            mm = '{:02}'.format(int(self.pd["timing"]["timecode"]["minutes"]))                
            ss = '{:02}'.format(int(self.pd["timing"]["timecode"]["seconds"]))
            ff = '{:02}'.format(int(self.pd["timing"]["timecode"]["frames"]))
            return hh + ':' + mm + ':' + ss + ':' + ff
        else:
            return None

    # Get the PTP sample time
    def Get_sample_time(self,part=None):
        if ("timing" in self.pd) and (self.pd["timing"]["sampleTimestamp"]):
            ssec = int(self.pd["timing"]["sampleTimestamp"]["seconds"])
            nsec = int(self.pd["timing"]["sampleTimestamp"]["nanoseconds"])
            asec = int(self.pd["timing"]["sampleTimestamp"]["attoseconds"])
            #ctime = ssec + (nsec * 0.000000001) + (asec * 0.000000000000000001)
            # Constants
            epoch = 1970                        # PTP is since this epoch
            spm = 60                            # seconds per minute. I know, but consistency is important
            sph = 3600                          # seconds per hour
            spd = 86400                         # sec per day
            spy = 31536000                      # sec per year
            # separate into years, days, hours, min, sec
            ydelta = int(ssec / spy)            # years since epoch
            yr = epoch + ydelta                 # current year
            sty = ssec - (ydelta * spy)         # seconds elapsed this year
            day = int(sty / spd)                # current day of year
            std = sty - (day * spd)             # seconds elapsed today (since midnight)
            hr = int(std / sph)                 # hours elapsed today
            mn = int((std - (hr * sph)) / spm)  # remainder minutes
            st = int(std - (hr * sph) - (mn * spm))  # remainder seconds
        if not part:
            if self.sample_time_format == "seconds":
                return ssec + (nsec * 0.000000001) + (asec * 0.000000000000000001)
            elif self.sample_time_format == "timecode":             # since midnight
                # FIXME: is this using the sample framerate?
                frm = int((nsec * 0.000000001) * self.Get_timecode_framerate())
                return '{:02}'.format(hr) + ":" + '{:02}'.format(mn) + ":" + '{:02}'.format(st) + ":" + '{:02}'.format(frm)
            elif self.sample_time_format == "string":
                return "year:{} day:{} hour:{} min:{} sec:{} nsec:{}".format(yr,day,hr,mn,st,nsec)
        else:
          pass  # FIXME: get year, day, hour, etc  

    # Frame rate which this timecode represents
    def Get_timecode_framerate(self):
        if ("timing" in self.pd) and (self.pd["timing"]["timecode"]["format"]["frameRate"]):
            numerator = float(self.pd["timing"]["timecode"]["format"]["frameRate"]["num"])  
            denominator = float(self.pd["timing"]["timecode"]["format"]["frameRate"]["denom"]) 
            return float(numerator / denominator)
        else:
            return None

    # Set user-preferred units for translations. 
    # Valid args: "m", "cm", "mm", "in"
    def Set_trans_units(self,unit_str):
        schema_units = self.sd["properties"]["transforms"]["items"]["items"]["properties"]["translation"]["units"]
        if self.verbose:
            print("Schema says camera translation units are {}".format(schema_units))
        print("Setting preferred translation units to: {0}".format(unit_str))
        if schema_units == "meters":
            self.trans_mult = self.Conversion_factor_from_meters(unit_str)

    # User preference for time format
    # Valid args: "timecode"
    def Set_sample_time_format(self,format_str):
        print("Setting preferred sample time format to: {0}".format(format_str))
        if (format_str in self.sample_time_formats):
            self.sample_time_format = format_str

    # Esablish a user-preference for units of focus distance.
    # Valid: "m","cm","mm","in"
    def Set_focus_distance_units(self,unit_str):
        schema_units = self.sd["properties"]["lens"]["properties"]["focusDistance"]["units"]
        if self.verbose:
            print("Schema says focus distance units are {}".format(schema_units))
        print("Setting preferred focus distance units to: {}".format(unit_str))
        if unit_str == "m":
            if schema_units == "millimeter":
                self.focus_dist_mult = .001
        elif unit_str == "cm":
            if schema_units == "millimeter":
                self.focus_dist_mult = 0.1
        elif unit_str == "mm":
            if schema_units == "millimeter":        
                self.focus_dist_mult = 1.0
        elif unit_str == "in":
            if schema_units == "millimeter":        
                self.focus_dist_mult = 1/25.4

    def Get_protocol(self):
        if (self.pd["protocol"]):
            return str(self.pd["protocol"])
        else:
            return None

    def Get_slate(self):
        if "device" in self.pd.keys():
            if self.pd["device"]["slate"]:
                return str(self.pd["device"]["slate"])
            else:
                return None
        else:
            return None

    # If present in this sample, the 'static' block has the active sensor dimensions
    def Get_sensor_dim_height(self):
        if "static" in self.pd: 
            if self.pd["static"]["camera"]["activeSensorResolution"]:
                height = int(self.pd["static"]["camera"]["activeSensorResolution"]["height"])
                return height
        return None

    # If present in this sample, the 'static' block has the active sensor dimensions
    def Get_sensor_dim_width(self):
        if "static" in self.pd: 
            if self.pd["static"]["camera"]["activeSensorResolution"]:
                width = int(self.pd["static"]["camera"]["activeSensorResolution"]["width"])
                return width
            else:
                return None
        return None

    # Find units in the schema for active sensor dimensions
    def Get_sensor_dim_units(self):
        if "static" in self.pd: 
            if self.sd["properties"]["camera"]["properties"]["activeSensorPhysicalDimensions"]:
                return str(self.sd["properties"]["camera"]["properties"]["activeSensorPhysicalDimensions"]["units"])
            else:
                return None
        else:
            return None

    # If present in this sample, the 'static' block would have the active sensor dimensions
    def Get_tracking_device_serial_number(self):
        if "static" in self.pd: 
            if self.pd["static"]["device"]["serialNumber"]:  # FIXME: helper function to do every level, Joseph write it?
                return str(self.pd["static"]["device"]["serialNumber"])
            else:
                return None
        return None

    def Get_focal_length(self):
        if "lens" in self.pd.keys():
            return self.pd["lens"]["focalLength"]
        else:
            return None

    def Get_focus_distance(self):
         return float(self.pd["lens"]["focusDistance"]) * self.focus_dist_mult

File: src/test/cpp/opentrackio-parser/opentrackio-parser/src/opentrackIOlib.cpp
// opentrackIOlib.cpp
//
// Library reference code for decoding opentrackIO messages
//
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the SMTPE RIS OSVP Metadata Project
//
// nlohmann JSON library licensed under the MIT License, Copyright (c) 2013-2022 Niels Lohmann

#include "opentrackIOlib.h"
#include <iostream>
#include <unordered_map>
#include <cmath>

#include <opentrackio-cpp/OpenTrackIOSample.h>
#include <nlohmann/json.hpp>

// constructor
void OTProtocol::Init(std::string msg_text, std::string schema_text, bool verbose){
    sample_str = msg_text; 
    schema_str = schema_text; 
    this->verbose = verbose; 
    trans_mult = 1.0; 
    rot_mult = 1.0; 
    sample_time_format = "sec"; 
    focus_dist_mult = 1.0;
 }
      
// Read the schema which governs the interpretation of the protocol
int OTProtocol::Import_schema(void) {
    if (!schema_str.empty()) {
        try {
            sd = nlohmann::json::parse(schema_str);
        } catch (nlohmann::json::parse_error&) {
            std::cout << "Got JSONDecodeError while decoding the sample!" << std::endl;
            sd = nullptr;
        }
        if (sd.is_null()) {
            std::cout << "Import_schema(): Failed to parse OpenTrackIO schema file." << std::endl;
        } else {
            std::cout << "Parsed the schema JSON successfully." << std::endl;
            if (verbose) {
                std::cout << "Contents of the parsed JSON schema dict:\n" << sd.dump(4) << "\n" << std::endl;
            }
        }
    } else {
        std::cout << "ERROR: no schema provided!" << std::endl;
        return -1;
    }
    return 0;
}

// Ingest the text and store the JSON items in a dictionary
int OTProtocol::Parse(void) {
    if (!sample_str.empty()) {
        std::cout << "Parsing JSON string from sample buffer..." << std::endl;
        if (!sample.initialise((std::string_view)sample_str)) {
            auto errors = sample.getErrors();
            for (auto error : errors) {
                std::cout << "Error: " << error << std::endl;
            }
            exit(-1);
        }
        if (verbose) {
            auto warnings = sample.getWarnings();
            for (auto warning : warnings) {
                std::cout << "Warning: " << warning << std::endl;
            }
            auto json = sample.getJson();
            std::cout << "Contents of the parsed JSON dict:\n" << json.dump(4) << "\n" << std::endl;
        }
    } else {
        std::cout << "Parse(): Error. No json text provided!" << std::endl;
        return -1;
    }
    return 0;
}

double Conversion_factor_from_meters(const std::string& unit_str) {
    if (unit_str == "m") {
        return 1.0;
    } else if (unit_str == "cm") {
        return 100.0;
    } else if (unit_str == "mm") {
        return 1000.0;
    } else if (unit_str == "in") {
        return (1000.0 / 25.4);
    }
    return 0.0;
}

double OTProtocol::Get_camera_trans(const std::string& dimension) {
    for (auto transform : sample.transforms->transforms) {
        if (transform.id == "Camera") {
            if (verbose) {
                std::cout << "found camera, dim = " << dimension << ", mult factor: " << trans_mult << std::endl;
            }
            if (dimension == "x") {
                return transform.translation.x * trans_mult;
            } else if (dimension == "y") {
                return transform.translation.y * trans_mult;
            } else if (dimension == "z") {
                return transform.translation.z * trans_mult;
            }
            break;
        }
    }
    return 0.0;
}

double OTProtocol::Get_camera_rot(const std::string& dimension) {
    for (auto transform : sample.transforms->transforms) {
        if (transform.id == "Camera") {
            if (dimension == "p") {
                return transform.rotation.pan * rot_mult;
            } else if (dimension == "t") {
                return transform.rotation.tilt * rot_mult;
            } else if (dimension == "r") {
                return transform.rotation.roll * rot_mult;
            }
            break;
        }
    }
    return 0.0;
}

// Return order: x,y,z
std::tuple<double, double, double> OTProtocol::Get_camera_translations(void) {
    return std::make_tuple(Get_camera_trans("x"), Get_camera_trans("y"), Get_camera_trans("z"));
}

// Return house timecode
std::string OTProtocol::Get_timecode(void) {
    std::string hh = std::to_string(sample.timing->timecode->hours);
    std::string mm = std::to_string(sample.timing->timecode->minutes);
    std::string ss = std::to_string(sample.timing->timecode->seconds);
    std::string ff = std::to_string(sample.timing->timecode->frames);
    return hh + ":" + mm + ":" + ss + ":" + ff;
}

// Get the PTP sample time
std::string OTProtocol::Get_sample_time(const std::string& part) {
    int ssec = sample.timing->sampleTimestamp->seconds;
    int nsec = sample.timing->sampleTimestamp->nanoseconds;
    // Constants
    const int epoch = 1970; // PTP is since this epoch
    const int spm = 60;     // seconds per minute. Common knowledge, but consistency is important
    const int sph = 3600;   // seconds per hour
    const int spd = 86400;  // sec per day
    const int spy = 31536000;   // sec per year
    // separate into years, days, hours, min, sec
    int ydelta = ssec / spy;            // years since epoch
    int yr = epoch + ydelta;            // current year
    int sty = ssec - (ydelta * spy);    // seconds elapsed this year
    int day = sty / spd;                // current day of year
    int std = sty - (day * spd);        // seconds elapsed today (since midnight)
    int hr = std / sph;                 // hours elapsed today
    int mn = (std - (hr * sph)) / spm;  // remainder minutes
    int st = std - (hr * sph) - (mn * spm); // remainder seconds

    if (part.empty()) {
        if (sample_time_format == "sec") {
            return std::to_string(ssec + (nsec * 0.000000001));
        } else if (sample_time_format == "timecode") {
            int frm = static_cast<int>((nsec * 0.000000001) * Get_timecode_framerate());
            return std::to_string(hr) + ":" + std::to_string(mn) + ":" + std::to_string(st) + ":" + std::to_string(frm);
        } else if (sample_time_format == "string") {
            return "year:" + std::to_string(yr) + " day:" + std::to_string(day) + " hour:" + std::to_string(hr) + 
                   " min:" + std::to_string(mn) + " sec:" + std::to_string(st) + " nsec:" + std::to_string(nsec);
        }
    } else {
        if (part == "yy") 
            return std::to_string(yr);
        else if (part == "dd") 
            return std::to_string(day);
        else if (part == "hh") 
            return std::to_string(hr);
        else if (part == "mm") 
            return std::to_string(mn);
        else if (part == "ss") 
            return std::to_string(st);
        else if (part == "ns") 
            return std::to_string(nsec);
    }
    return "";
}

// Frame rate which this timecode represents
double OTProtocol::Get_timecode_framerate(void) {
    double numerator = sample.timing->timecode->format.frameRate.numerator;
    double denominator = sample.timing->timecode->format.frameRate.denominator;
    return numerator / denominator;
}

// Set user-preferred units for translations. 
// Valid args: "m", "cm", "mm", "in"
void OTProtocol::Set_trans_units(const std::string& unit_str) {
    std::string schema_units = sd["properties"]["transforms"]["items"]["properties"]["translation"]["units"];
    if (verbose) {
        std::cout << "Schema says camera translation units are: " << schema_units << std::endl;
        std::cout << "Setting preferred translation units to: " << unit_str << std::endl;
    }
    if (schema_units == "meter") {
        trans_mult = Conversion_factor_from_meters(unit_str);
    }
}

// Set user-preferred units for rotations. 
// Valid args: "deg", "rad"
void OTProtocol::Set_rotation_units(const std::string& unit_str) {
    std::string schema_units = sd["properties"]["transforms"]["items"]["properties"]["rotation"]["units"];
    if (verbose) {
        std::cout << "Schema says camera rotation units are: " << schema_units << std::endl;
        std::cout << "Setting preferred camera rotation units to: " << unit_str << std::endl;
    }
    if (schema_units == "degree") {
        if (unit_str == "deg") {
            rot_mult = 1.0;
        } else if (unit_str == "rad") {
            rot_mult = M_PI / 180;
        }
    }
}

// User preference for time format
// Valid args: "sec", "timecode", "string"
void OTProtocol::Set_sample_time_format(const std::string& format_str) {
    if (verbose) {
        std::string schema_units = sd["properties"]["timing"]["properties"]["sampleTimestamp"]["units"];
        std::cout << "Schema says sample time units are: " << schema_units << std::endl;
        std::cout << "Setting preferred sample time format to: " << format_str << std::endl;
    }
    if (std::find(sample_time_formats.begin(), sample_time_formats.end(), format_str) != sample_time_formats.end()) {
        sample_time_format = format_str;
    }
}

// Establish a user-preference for units of focus distance by storing a conversion factor.
// Valid: "m","cm","mm","in"
void OTProtocol::Set_focus_distance_units(const std::string& unit_str) {
    std::string schema_units = sd["properties"]["lens"]["properties"]["focusDistance"]["units"];
    if (verbose) {
        std::cout << "Schema says focus distance units are: " << schema_units << std::endl;
        std::cout << "Setting preferred focus distance units to: " << unit_str << std::endl;
    }
    if (schema_units == "millimeter") {
        if (unit_str == "m") {
            focus_dist_mult = 0.001;
        } else if (unit_str == "cm") {
            focus_dist_mult = 0.1;
        } else if (unit_str == "mm") {
            focus_dist_mult = 1.0;
        } else if (unit_str == "in") {
            focus_dist_mult = 1.0 / 25.4;
        }
    }
}

// The protocol to which this sample conforms
std::string OTProtocol::Get_protocol() {
    return sample.protocol->name + " v" + sample.protocol->version;
}

std::string OTProtocol::Get_slate(void) {
    if (sample.tracker->slate) {
        return *sample.tracker->slate;
    }
    return "";
}

// If present in this sample, the 'static' block has the active sensor dimensions
int OTProtocol::Get_sensor_dim_height(void) {
    if (sample.camera->activeSensorResolution->height) {
        return sample.camera->activeSensorResolution->height;
    }
    return 0;
}

// If present in this sample, the 'static' block has the active sensor dimensions
int OTProtocol::Get_sensor_dim_width(void) {
    if (sample.camera->activeSensorResolution->width) {
        return sample.camera->activeSensorResolution->width;
    }
    return 0;
}

std::string OTProtocol::Get_sensor_dim_units(void) {
    if (sd["properties"]["camera"].contains("activeSensorPhysicalDimensions")) {
        return sd["properties"]["camera"]["activeSensorPhysicalDimensions"]["units"].get<std::string>();
    }
    return "";
}

std::string OTProtocol::Get_tracking_device_serial_number(void) {
    if (sample.tracker->serialNumber) {
        return *sample.tracker->serialNumber;
    }
    return "";
}

double OTProtocol::Get_focal_length(void) {
    if (sample.lens->focalLength) {
        return *sample.lens->focalLength;
    }
    return 0.0;
}

double OTProtocol::Get_focus_distance(void) {
    if (sample.lens->focusDistance) {
        return *sample.lens->focusDistance * focus_dist_mult;
    }
    return 0.0;
}
